### PR TITLE
Alexis jorge push database

### LINKED ISSUE #13: Implement handler to store data from gRPC push in database
Sensor -> Server

By Sunday

### PR DESCRIPTION
**Pull request by penguingovernor**
Should close #13 Might conflict with [this](https://github.com/UCSC-CSE123/beavertail/pull/17)

### CODE REVIEW

**Review by LexiKat — APPROVED** [review]

**Review by LexiKat — APPROVED** [review]

File: server/main.go [hunk 1]
```diff
@@ -1,27 +1,32 @@
...
+	// and init. the databases associated with it.
+	srv := datagram.NewServer(db)
+	if err := srv.InitDatabase(); err != nil {
+		log.Fatalf("failed to init. db: %s\n", err)
+	}
+
+	// Start the server.
+	const host = "localhost"
+	const port = "3000"
+	if err := srv.Serve(host, port); err != nil {
```
Comment (natanlao) [inline server/main.go]: Can you add documentation on how to run this? I assume it's `go run server/main.go`, but it would be good to have that written down somewhere for clarity.

File: server/models/db.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import "database/sql"
+
+// NewDB instantiates a new connection to the sqllite3 server.
+func NewDB(dataSource string) (*sql.DB, error) {
+	db, err := sql.Open("sqlite3", dataSource)
+	if err != nil {
+		return nil, err
+	}
+	if err = db.Ping(); err != nil {
```
Comment (natanlao) [inline server/models/db.go:12]: Good call.

File: server/models/db.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Package models defines all data models for the beavertail server.
+package models
+
+import "database/sql"
+
+// NewDB instantiates a new connection to the sqllite3 server.
```
Comment (natanlao) [inline server/models/db.go]: Nit
```suggestion
// NewDB instantiates a new connection to the sqlite3 database.
```

File: server/main.go [hunk 4]
```diff
@@ -1,27 +1,32 @@
...
+
+	"github.com/UCSC-CSE123/beavertail/server/models"
+	"github.com/UCSC-CSE123/beavertail/server/service/datagram"
+	_ "github.com/mattn/go-sqlite3"
 )
 
 func main() {
-	db, err := sql.Open("sqlite3", "./busdata.db")
+	// Establish a connection to the database.
+	db, err := models.NewDB("./busdata.db")
```
Comment (natanlao) [inline server/main.go]: Can you make the database path configurable by argument?

File: server/service/datagram/serve.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"net"
+
+	pb "github.com/UCSC-CSE123/beavertail/server/beavertail"
+	"google.golang.org/grpc"
+)
+
+func (srv *Server) Serve(host, port string) error {
+	// Setup a lister.
```
Comment (natanlao) [inline server/service/datagram/serve.go]:
```suggestion
	// Setup a listener.
```

File: server/main.go [hunk 6]
```diff
@@ -1,27 +1,32 @@
 package main
 
 import (
-	"database/sql"
-	_ "github.com/mattn/go-sqlite3"
 	"log"
+
+	"github.com/UCSC-CSE123/beavertail/server/models"
```
Comment (natanlao) [inline server/main.go:7]: Is there a reason that this doesn't reference source files locally? (This isn't a leading question - I'm not sure if this is best practice or not.)

File: server/service/datagram/push.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	// If there was no error,
+	// return a point to a datagram ack.
+	return &pb.DatagramAck{
+		Acknowledgment: pb.DatagramAck_OK,
+	}, nil
+}
+
+// InitDatabase initializes the database associated with this server.
+func (srv *Server) InitDatabase() error {
```
Comment (natanlao) [inline server/service/datagram/push.go]: Why is this here and not in `server/main.go`? Also, is there a reason that `InitDatabase` and `NewDB` are separate?

**Review by natanlao — CHANGES_REQUESTED** [review]
Thanks for working on this! This looks good overall, just a few surface-level tweaks and questions.

File: server/main.go (hunk 6 above)
Comment (penguingovernor) [reply server/main.go:7]: Since our module name is `github.com/UCSC-CSE123/beavertail/server/` we'll have to reference any packages we make in it as `moduleName/pkgName`. Relative imports like `../pkgName` are a heated conversation in the Go community, and it seems like absolute imports are the way to go for the time being.

File: server/service/datagram/push.go (hunk 7 above)
Comment (penguingovernor) [reply server/service/datagram/push.go]: I'd like to keep `server/main.go` as lean as possible and try to have handle only the command line items there. `NewDB` is seperate from `InitDatabase` because `NewDB` just opens a new connection to the database and make sures its okay, whereas `InitDatabase` specifically inits the database for our datagram service, thus the reason it's in the `server/service/datagram`. 

I'm happy to make any changes though (:

File: server/main.go (hunk 4 above)
Comment (penguingovernor) [reply server/main.go]: Paths are now [con]figurable via flags or a yaml file (: 

edit: typos

File: server/main.go (hunk 6 above)
Comment (natanlao) [reply server/main.go:7]: Interesting, I didn't know there was contention.

File: server/service/datagram/push.go (hunk 7 above)
Comment (natanlao) [reply server/service/datagram/push.go]: For sure. Intuitively, I don't see a reason for them to be separated, since both functions are called unconditionally in `server/main.go`.

**Review by natanlao — APPROVED** [review]

**Comment by natanlao** [comment]
Thanks for addressing changes! This looks fine to me, [this comment](https://github.com/UCSC-CSE123/beavertail/pull/19#discussion_r412602718) notwithstanding. That being said, I think you have a better idea of the project architecture than I do at this point, so I trust your judgement.

File: server/service/datagram/push.go (hunk 7 above)
Comment (penguingovernor) [reply server/service/datagram/push.go]: After more thought I think your reasoning is correct, because if we do add more servers hypothetically, they would all be initing the database and I'm pretty sure thats no bueno. I'll seperate these and merge (:
